src/mlpack/methods/ann/layer/dense.hpp

```diff
@@ -0,0 +1,197 @@
+#include "convolution.hpp"
+#include "dropout.hpp"
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The dense block is a layer which connects each layer
+ * to every other layer in a feedforward fashion. Compared to
+ * traditional neural networks with L connections (connecting
+ * the subsequent layers), the dense block has L(L+1)/2 connections.
+ * It alleviates the vanishing-gradient problem and strengthens
+ * feature propagation. It also encourages feature reuse and
+ * substantially reduces the number of parameters.
+ * 
+ * For more information, see the following.
+ * 
+ * @article{DBLP:journals/corr/HuangLW16a,
+ * author    = {Gao Huang and
+ *              Zhuang Liu and
+ *              Kilian Q. Weinberger},
+ * title     = {Densely Connected Convolutional Networks},
+ * journal   = {CoRR},
+ * volume    = {abs/1608.06993},
+ * year      = {2016},
+ * url       = {http://arxiv.org/abs/1608.06993},
+ * archivePrefix = {arXiv},
+ * eprint    = {1608.06993},
+ * timestamp = {Mon, 10 Sep 2018 15:49:32 +0200},
+ * biburl    = {https://dblp.org/rec/bib/journals/corr/HuangLW16a},
+ * bibsource = {dblp computer science bibliography, https://dblp.org}
+ * }
+ */
+template<typename InputDataType = arma::cube,
+         typename OutputDataType = arma::cube>
+class Dense
+{
+ public:
+  //! Create the Dense object.
+  Dense();
+
+  /**
+   * Create the Dense layer object for a specific configuration.
+   * 
+   * @param nb_layers The number of convolution blocks to append to the block.
+   * @param growth_rate Growth rate of the dense block.
+   * @param bottleneck Bottleneck convolution block is added to each convolution
+   *        block if true.
+   * @param dropout_rate Dropout rate for the dropout layer.
+   * @param weight_decay Weight decay factor.
+   */
+  Dense(const size_t nb_layers, const size_t growth_rate,
+    const bool bottleneck = false, const double dropout_rate = 0.2,
+    const double weight_decay = 1e-4);
+
+  /**
+   * Create the Convolution block used for building dense blocks.
+   * 
+   * @param input Input data used for evaluating the specified function.
+   * @param growth_rate Growth rate of the dense block.
+   * @param bottleneck Bottleneck convolution block is added to each convolution
+   *        block if true.
+   * @param dropout_rate Dropout rate for the dropout layer.
+   * @param weight_decay Weight decay factor.
+   */
+  template<typename eT>
+  arma::cube conv_block(arma::Cube<eT>&& input, const size_t& input_size,
+    const size_t& growth_rate, const bool& bottleneck,
+    const double& dropout_rate, const double& weight_decay);
+
+  /**
+   * Ordinary feed forward pass of the dense layer.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param output Resulting output activation.
+   */
+  template<typename eT>
+  void Forward(const arma::Cube<eT>&& input, arma::Cube<eT>&& output);
```

Comment:
Since the visitor (e.g. parameter) does expect ``arma::Mat`` instead of ``arma::Cube`` it returns with a bunc hof errors, so either we add the ability to handle ``arma::Cube`` as well or we use ``arma::Mat`` here and use the advanced constructor of ``arma::Cube`` to transform ``arma::Mat`` into a ``arma::Cube``. Let me know what you think.

Reply:
I'm not sure if I understood the visitor part, are you talking about the files in `ann/visitor`? I feel that it would be better to add the ability to handle `arma::Cube` as it will be useful for other future implementations as well. Building an advanced constructor here might lead to code redundancies if some other layer also wants to use `arma::Cube`. I'm not sure if I'm correct :sweat_smile: , please let me know what you think :)

Reply:
Right I was talking about ``ann/visitor``, if we go with the ``ann/visitor`` adaption we have to think about a way to pass between ``arma::cube`` and ``arma::mat``, e.g. a linear layer has to pass the output to the next layer that will take an ``arma::cube`` as input.

Reply:
Yes :+1: I checked the files `add_visitor.hpp` and `add_visitor_impl.hpp` but could not find `arma::mat` explicitly used anywhere (All the functions are using templates); am I missing something? I checked `ffn.hpp` and it has it's own `Add()`... What exactly is the purpose of the `visitor` classes?

Reply:
The visitor paradigma enables us to call different functions on ``boost::variant`` the data structure that is used to hold the layer.

So, since the visitor classes do work on templates, I think the only class we have to adapt is FFN and RNN, since they internally expect that the input/output parameter for each layer doesn't change.

Perhaps it makes more sense to use ``arma::mat`` instead of ``arma::cube`` for the Dense layer as well.

Reply:
How can I concatenate layers in `arma::mat`? The reason I used `arma::cube` is that it had the `join_slices()` function which I couldn't find for the `arma::mat` class. Did I miss some function?

Reply:
Sorry for the slow response, you could interpret the cols as slice, you can also transform an ``arma::mat`` into an ``arma::cube`` if you like to use ``join_slices()``:

```c++
arma::cube transformedMat = arma::cube(mat.memptr(), rows, cols, slices, false, false);
```

Reply:
> Sorry for the slow response, you could interpret the cols as slice, you can also transform an `arma::mat` into an `arma::cube` if you like to use `join_slices()`:
> 
> ```c++
> arma::cube transformedMat = arma::cube(mat.memptr(), rows, cols, slices, false, false);
> ```

:+1: I'll try this method and let you know.